client/rabbitmq.go

```diff
@@ -97,91 +133,90 @@ func (c *RabbitClient) Call(ctx context.Context, serviceName, endpoint string, r
+		log.Debugf("[Client] Response received for %s from %s", req.Id(), routingKey)
+		rsp := deliveryToResponse(delivery)
+		if rsp.IsError() {
+			return nil, unmarshalErrorResponse(rsp)
+		}
+		return rsp, nil
 	case <-time.After(defaultTimeout):
-		log.Errorf("[Client] Request %s timed out calling %s", correlation.String(), routingKey)
+		log.Errorf("[Client] Request %s timed out calling %s", req.Id(), routingKey)
 
-		return errors.Timeout(fmt.Sprintf("%s timed out", routingKey), nil, map[string]string{
-			"called_service":  serviceName,
-			"called_endpoint": endpoint,
+		return nil, errors.Timeout(fmt.Sprintf("%s timed out", routingKey), nil, map[string]string{
+			"called_service":  req.Service(),
+			"called_endpoint": req.Endpoint(),
 		})
 	}
+
 }
 
-func (c *RabbitClient) buildRoutingKey(serviceName, endpoint string) string {
+// buildRoutingKey to send the request via AMQP
+func buildRoutingKey(serviceName, endpoint string) string {
 	return fmt.Sprintf("%s.%s", serviceName, endpoint)
 }
 
-// handleResponse returned from a service by marshaling into the response type,
-// or converting an error from the remote service
-func handleResponse(delivery amqp.Delivery, resp proto.Message) error {
-	// deal with error responses, by converting back from wire format
-	if deliveryIsError(delivery) {
-		p := &pe.Error{}
-		if err := proto.Unmarshal(delivery.Body, p); err != nil {
-			return errors.BadResponse(err.Error())
-		}
-
-		return errors.Unmarshal(p)
-	}
-
-	// Otherwise try to marshal to the expected response type
-	if err := proto.Unmarshal(delivery.Body, resp); err != nil {
+// unmarshalResponse returned from a service into the response type
+func unmarshalResponse(resp Response, respProto proto.Message) error {
+	if err := proto.Unmarshal(resp.Payload(), respProto); err != nil {
 		return errors.BadResponse(err.Error())
 	}
 
 	return nil
 }
 
-// deliveryIsError checks if the delivered response contains an error
-func deliveryIsError(delivery amqp.Delivery) bool {
-	encoding, ok := delivery.Headers["Content-Encoding"].(string)
-	if !ok {
-		// Can't type assert header to string, assume error
-		log.Warnf("[Client] Service returned invalid Content-Encoding header %v", encoding)
-		return true
+// deliveryToResponse converts our AMQP response to a client Response
+func deliveryToResponse(delivery amqp.Delivery) Response {
+
+	contentType, _ := delivery.Headers["Content-Type"].(string)
+	contentEncoding, _ := delivery.Headers["Content-Encoding"].(string)
+	service, _ := delivery.Headers["Service"].(string)
+	endpoint, _ := delivery.Headers["Endpoint"].(string)
+
+	return &response{
+		contentType:     contentType,
+		contentEncoding: contentEncoding,
+		service:         service,
+		endpoint:        endpoint,
+		payload:         delivery.Body,
 	}
+}
 
-	if encoding == "" || encoding == "ERROR" {
-		return true
+// unmarshalErrorResponse from our wire format to a typhon error
+func unmarshalErrorResponse(resp Response) *errors.Error {
```

Comment:
Unfortunately, even in this special case, we should return an `error` interface. Otherwise the line `errors.Unmarshal(p)` may return a "corrupted" `nil` that we can never return from a method as an `error` interface